### PR TITLE
Phase 6: EMA Stochastic Weight Perturbation — Flat Minima Seeking

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Models in flatter loss basins generalize better to distribution shifts. p_tan tests on NACA6416 (unseen foil) — the sharpest OOD challenge. The current EMA (decay=0.999, starting at epoch ~140) averages the training trajectory but doesn't explicitly encourage flat-basin convergence.

**Idea:** At the moment EMA starts accumulating, inject a one-time Gaussian perturbation into the live model weights. The model then recovers from this perturbation over subsequent training steps, and the EMA averages the recovery trajectory. If the loss basin is flat (good for generalization), the recovery is smooth and the EMA captures a broader region of the flat manifold. If the basin is sharp, recovery is rocky and EMA captures worse models.

This is inspired by Stochastic Weight Averaging (Izmailov et al., 2018) and the "perturbation → recovery → average" principle, but simpler than full SAM (which is incompatible with Lion, PR #2086).

**Key difference from prior failed methods:**
- SAM (#2086): per-step perturbation, incompatible with Lion's sign operation
- SGLD (#2120): per-step noise, overwhelmed by Lion's implicit gradient noise
- SWAD (#2094): dense trajectory averaging over many checkpoints, catastrophic
- **This:** ONE perturbation at EMA start, then normal training + EMA. No per-step overhead. No interaction with Lion optimizer.

**Expected impact:** -1% to -3% p_tan. Low risk — worst case the perturbation is too small to matter.

### Step 1: Add config flag

In `Config` dataclass:
```python
ema_perturb_sigma: float = 0.0   # One-time weight perturbation magnitude at EMA start (0 = disabled)
```

In argparse:
```python
parser.add_argument('--ema_perturb_sigma', type=float, default=0.0,
                    help='Gaussian perturbation sigma applied to model weights at EMA start epoch')
```

### Step 2: Add perturbation at EMA initialization

Find the EMA initialization point in the training loop (search for `ema_model is None` or `deepcopy`, around line 1928-1931). It looks like:

```python
if epoch >= cfg.ema_start_epoch and not cfg.swad and ...:
    if ema_model is None:
        ema_model = deepcopy(_base_model)
    else:
        # existing EMA update ...
```

Add the perturbation BEFORE the deepcopy:

```python
if epoch >= cfg.ema_start_epoch and not cfg.swad and ...:
    if ema_model is None:
        # One-time weight perturbation to explore flat minima
        if cfg.ema_perturb_sigma > 0.0:
            with torch.no_grad():
                for p in _base_model.parameters():
                    if p.requires_grad:
                        p.data.add_(torch.randn_like(p.data) * cfg.ema_perturb_sigma)
            print(f"[ema_perturb] Applied sigma={cfg.ema_perturb_sigma} perturbation at epoch {epoch}")
        ema_model = deepcopy(_base_model)
    else:
        # existing EMA update ...
```

**Important:** The perturbation modifies the LIVE model, not the EMA. The EMA starts from the perturbed model and then averages the recovery trajectory.

### Step 3: Run sweep — 3 sigma values × 2 seeds = 6 runs

```bash
BASE="--asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context"

# sigma=5e-4, seeds 42/73
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-5e4-s42" \
  --wandb_group phase6/ema-perturb --seed 42 --ema_perturb_sigma 5e-4 $BASE
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-5e4-s73" \
  --wandb_group phase6/ema-perturb --seed 73 --ema_perturb_sigma 5e-4 $BASE

# sigma=1e-3, seeds 42/73
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-1e3-s42" \
  --wandb_group phase6/ema-perturb --seed 42 --ema_perturb_sigma 1e-3 $BASE
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-1e3-s73" \
  --wandb_group phase6/ema-perturb --seed 73 --ema_perturb_sigma 1e-3 $BASE

# sigma=3e-3, seeds 42/73
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-3e3-s42" \
  --wandb_group phase6/ema-perturb --seed 42 --ema_perturb_sigma 3e-3 $BASE
python train.py --agent tanjiro --wandb_name "tanjiro/ema-perturb-3e3-s73" \
  --wandb_group phase6/ema-perturb --seed 73 --ema_perturb_sigma 3e-3 $BASE
```

Run all 6 in parallel. W&B group: `phase6/ema-perturb`.

### What to report

1. Confirm the perturbation print fired at the expected epoch
2. Surface MAE table (p_in, p_oodc, p_tan, p_re) for all 6 runs + W&B run IDs
3. 2-seed means per sigma vs baseline
4. val/loss comparison — did the perturbation cause a visible spike followed by recovery?

## Baseline

Current single-model baseline (PR #2127 — AftSRF KNN Context K=8):

| Metric | 2-seed avg | Target |
|--------|------------|--------|
| p_in | **13.02** | < 13.02 |
| p_oodc | **7.62** | < 7.62 |
| **p_tan** | **29.91** | **< 29.91** |
| p_re | **6.47** | < 6.47 |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-aft-srf-ctx" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Implementation

Added `ema_perturb_sigma` config flag and one-time Gaussian weight perturbation at EMA initialization (epoch 140). Perturbation is applied to all model components (main model, refinement head, aft-foil SRF head, aft-foil SRF context head) before deepcopy. Confirmed `[ema_perturb]` message fired at epoch 140 for all runs.

**Note:** First batch of 5/6 runs had a shell variable expansion bug (inline `$BASE` wasn't propagated to background subshells), so they ran without critical flags like `--surface_refine`. Re-ran those 5 with fully inlined args. Only `run_5e4_s42` from the first batch was valid.

### Results Table

| Run | σ | Seed | Best Epoch | val/loss | p_in | p_tan | p_oodc | p_re | W&B |
|-----|---|------|-----------|----------|------|-------|--------|------|-----|
| 5e4-s42 | 5e-4 | 42 | 156 | 0.3851 | 13.1 | **29.8** | 7.7 | 6.5 | `er41ypz5` |
| 5e4-s73 | 5e-4 | 73 | 157 | 0.3916 | 12.7 | 31.2 | 7.5 | 6.5 | `asbtprv1` |
| 1e3-s42 | 1e-3 | 42 | 157 | 0.3849 | 13.0 | **29.8** | 7.7 | 6.5 | `obkrg47t` |
| 1e3-s73 | 1e-3 | 73 | 156 | 0.3908 | 13.4 | 30.4 | 7.8 | 6.5 | `6s1r2mu7` |
| 3e3-s42 | 3e-3 | 42 | 157 | 0.3857 | 13.1 | **29.9** | 7.7 | 6.5 | `4w1o1rnr` |
| 3e3-s73 | 3e-3 | 73 | 157 | 0.3911 | 13.3 | 30.5 | 7.7 | 6.5 | `2nxtud27` |
| **Baseline** | — | — | — | — | **13.02** | **29.91** | **7.62** | **6.47** | — |

### 2-Seed Averages vs Baseline

| σ | p_in | p_tan | p_oodc | p_re |
|---|------|-------|--------|------|
| 5e-4 | 12.9 (-0.9%) | **30.5 (+2.0%)** | 7.6 (-0.3%) | 6.5 (+0.5%) |
| 1e-3 | 13.2 (+1.4%) | **30.1 (+0.6%)** | 7.75 (+1.7%) | 6.5 (+0.5%) |
| 3e-3 | 13.2 (+1.4%) | **30.2 (+1.0%)** | 7.7 (+1.0%) | 6.5 (+0.5%) |

### Command Used

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=<gpu> python train.py --agent tanjiro \
  --wandb_name "tanjiro/ema-perturb-<sigma>-s<seed>" --wandb_group phase6/ema-perturb \
  --seed <seed> --ema_perturb_sigma <sigma> \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

Peak memory: ~38 GB (unchanged from baseline).

### Analysis

**The perturbation did not improve p_tan at any sigma.** All 2-seed averages are above the baseline 29.91:

- **σ=5e-4:** Avg p_tan=30.5 (+2.0%). Seed 42 was competitive (29.8) but seed 73 regressed badly (31.2). High inter-seed variance suggests the perturbation adds noise rather than systematic improvement.
- **σ=1e-3:** Avg p_tan=30.1 (+0.6%). Closest to baseline, but seed 42 (29.8) matches baseline-level while seed 73 (30.4) regresses.
- **σ=3e-3:** Avg p_tan=30.2 (+1.0%). Similar pattern — seed 42 reasonable, seed 73 regresses.

**Key observation:** Seed 42 consistently outperforms seed 73 across ALL sigma values (29.8-29.9 vs 30.4-31.2). This suggests the perturbation's effect is dominated by seed-dependent basin geometry, not by systematic flat-minima seeking.

**Why it didn't work:**
1. **Too little recovery time:** EMA starts at epoch 140, training ends at epoch ~157. Only ~17 epochs of recovery + EMA averaging — not enough for the model to traverse and average a meaningful portion of the flat basin.
2. **Lion already provides implicit exploration:** Lion's sign operation discards gradient magnitude, and with ema_decay=0.999, the EMA is already capturing a very smooth average of the training trajectory.
3. **Cosine LR near minimum at epoch 140:** The learning rate at epoch 140/160 is near its minimum, so recovery from perturbation is very slow. The model barely moves from the perturbed position.
4. **Global sigma is too crude:** A fixed sigma applied uniformly to all parameters doesn't account for scale differences across layers. The perturbation may be too large for some parameters and too small for others.

### Suggested Follow-ups

1. **Perturbation-relative-to-weight-norm (σ * ||w||):** Scale perturbation by parameter magnitude so each layer gets proportionally sized noise. This addresses the "one sigma doesn't fit all" problem.
2. **Earlier perturbation (epoch 100-120) with more recovery time:** Move EMA start earlier and perturb then, giving 40-50 epochs of recovery instead of 17.
3. **Cyclical LR restart at perturbation:** Bump the learning rate back up when the perturbation is applied, mimicking SWA's cyclical schedule. This would allow the model to actually explore and recover meaningfully.
4. **Given that SAM, SGLD, SWAD, and now single-shot perturbation all failed to beat baseline, flat-minima-seeking approaches may have limited ROI for this architecture+optimizer combination.** The Lion + cosine schedule + EMA combination may already be a sufficient implicit regularizer.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2137 because ADVISOR: Closing — EMA weight perturbation is a dead end. All σ values regress p_tan vs old baseline (30.5, 30.1, 30.2 vs 29.91), and these runs are far from the current baseline (28.60).

Your analysis point #4 is spot-on: **flat-minima-seeking is exhausted for this setup.** SAM (#2086), SGLD (#2120), SWAD (#2094), and now EMA perturbation have all failed. The Lion + cosine + EMA combination already provides sufficient implicit regularization. This is a confirmed dead end CLASS, not just individual experiments.

The seed 42 vs 73 divergence (consistently 0.6-1.4 p_tan points) reinforces that p_tan is high-variance and 2-seed averages are essential for reliable signals.

**Note:** Runs used stale baseline flags (--disable_pcgrad, --aft_foil_srf_context, no GSB). Results not comparable to current baseline.

New assignment coming shortly..
